portfolio/src/test/java/com/google/sps/BookDaoTest.java

```diff
@@ -0,0 +1,214 @@
+    ArrayList<String> authors = new ArrayList<String>();
+    authors.add(author);
+    ArrayList<String> categories = new ArrayList<String>();
+    categories.add(category);
+    Set<String> genres = new HashSet<String>();
+    genres.add(genre);
+    Book.Builder builder = Book.builder().title(title).authors(authors)
+            .categories(categories).language(language)
+            .genre(genres).pageCount(pageCount).isbn(isbn);
+    Book expected = builder.build();
+    return expected;
+  }
+  
+  @Test(expected = NullPointerException.class)
+  public void validateNullIsbn(){
+    BookDao.getEntity(null);
+  }
+  
+  @Test(expected = RuntimeException.class)
+  public void validateEmptyIsbn(){
+    BookDao.getEntity("");
+  }
+  
+  @Test(expected = RuntimeException.class)
+  public void validateShortIsbn(){
+    BookDao.getEntity("1234567");
+  }
+  
+  @Test(expected = RuntimeException.class)
+  public void validateLongIsbn(){
+    BookDao.getEntity("1234567891124151234");
+  }
+  
+  @Test(expected = RuntimeException.class)
+  public void validateNonNumericIsbn(){
+    BookDao.getEntity("puppies");
+  }
+  
+  @Test(expected = RuntimeException.class)
+  public void validateIllegalCharIsbn(){
+    BookDao.getEntity("1+2-344444444");
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void validateBookNullIsbn(){
+    Book nullIsbn = newBook("A Court of Wings and Ruin","Fiction","Sarah Maas","Fantasy","en",34,null);
+    BookDao.create(nullIsbn);
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void validateBookEmptyIsbn(){
+    Book emptyIsbn = newBook("A Court of Wings and Ruin","Fiction","Sarah Maas","Fantasy","en",34,"");
+    BookDao.create(emptyIsbn);
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void validateBookShortIsbn(){
+    Book shortIsbn = newBook("A Court of Wings and Ruin","Fiction","Sarah Maas","Fantasy","en",34,"12345");
+    BookDao.create(shortIsbn);
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void validateBookLongIsbn(){
+    Book longIsbn = newBook("A Court of Wings and Ruin","Fiction","Sarah Maas","Fantasy","en",34,"1234543243792349374074");
+    BookDao.create(longIsbn);
+  }
+
+  @Test
+  public void createAndGetBook() {
+    //creates a book and stores in DS, gets the book and compares
+    BookDao.create(defaultBook);
+    Book actual = BookDao.getEntity(defaultBook.isbn());
+    compareBooks(defaultBook,actual);
+  }
+
+  @Test
+  public void createExistingBook() {
+    //tries to create a book that is already in datastore
+    BookDao.create(defaultBook);
+  }
+
+  @Test
+  public void updateBook() {
+    //updates a book and in DS, gets the book and compares to original
+    Book original = newBook("A Court of Wings and Ruin","Fiction","Sarah Maas","Fantasy","en",34,"9781619634497");
+    BookDao.create(original);
+
+    Book updatedBook = newBook("A Court of Ruin and Wings","Fiction","Sarah Maas","Fantasy","en",34,"9781619634497");
```

Comment:
You should make some changes in teh updated book, otherwise it's hard to tell if the returned book is original or updated.

Reply:
I changed the name of the book from "A Court of Wings and Ruin" to "A Court of Ruin and Wings"

Reply:
Ah, I didn't see that, my bad :)